PostHog/Replay/NetworkRequestSample.swift

```diff
@@ -0,0 +1,257 @@
+    import Foundation
+
+    private let jsonRegex: String = "^application/.*json"
+    private let xmlRegex: String = "^application/.*xml"
+
+    class NetworkRequestSample: Identifiable {
+        enum ContentType: String {
+            case json
+            case xml
+            case html
+            case image
+            case other
+
+            init(contentType: String) {
+                switch contentType {
+                case _ where contentType.matches(jsonRegex):
+                    self = .json
+                case _ where contentType.matches(xmlRegex) || contentType == "text/xml":
+                    self = .xml
+                case "text/html":
+                    self = .html
+                case _ where contentType.hasPrefix("image/"):
+                    self = .image
+                default:
+                    self = .other
+                }
+            }
+        }
+
+        lazy var id = UUID().uuidString
+
+        var timestamp = getCurrentTimeMilliseconds()
+        var timeOrigin = getMonotonicTimeInMilliseconds()
+
+        var requestStartTime: UInt64?
+        var requestURL: URL?
+        var requestMethod: String?
+        var requestHeaders: [String: Any]?
+        var requestContentType: ContentType?
+        var requestContentTypeRaw: String?
+        var requestBodyStr: String?
+        var requestBodyLength: Int?
+
+        var responseError: String?
+        var responseData: NSMutableData?
+        var responseStatus: Int?
+        var responseContentType: ContentType?
+        var responseContentTypeRaw: String?
+        var responseStartTime: UInt64?
+        var responseEndTime: UInt64?
+        var responseHeaders: [String: Any]?
+        var responseBodyStr: String?
+        var responseBodyLength: Int?
+
+        var durationMs: UInt64?
+
+        var isProcessed: Bool = false
+
+        // called when a request starts loading
+        func start(request: URLRequest) {
+            requestStartTime = getMonotonicTimeInMilliseconds()
+            requestURL = request.url?.absoluteURL
+            requestMethod = request.httpMethod
+            requestHeaders = request.normalizedHeaderFields ?? [:]
+
+            // grab content-type. Keys normalized with .lowercase()
+            if let contentType = requestHeaders?["content-type"] as? String {
+                let contentType = contentType.components(separatedBy: ";")[0]
+                requestContentTypeRaw = contentType
+                requestContentType = ContentType(contentType: contentType)
+            }
+
+            // grab request body
+            if let requestData = request.httpBody ?? request.httpBodyStream?.consume() {
+                if let responseContentType, responseContentType == .image {
+                    // don't record response body for image types
+                    let bodyStr = requestData.base64EncodedString(options: .endLineWithLineFeed)
+                    requestBodyLength = bodyStr.count
```

Comment:
body length in bytes is != than base64 encoded char count.
we can probably read `Content-Length` here or just count bytes from `requestData`